### PR TITLE
Add service worker plugin

### PR DESCRIPTION
**Pull request by mwistrand**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

**Description:**

This is the `webpack-contrib` counterpart to dojo/cli-build-app#96. The `ServiceWorkerPlugin` will either generate a service worker from the provided config, or simply ensure an existing service worker file is copied to the output directory. It also provides a custom entry point that can be added to a webpack config to register the generated service worker so that app developers are not required to do so manually.

This plugin uses workbox under the hood via the [workbox-webpack-plugin](https://www.npmjs.com/package/workbox-webpack-plugin), but while the `ServiceWorkerPlugin` supports nearly all the options supported by `workbox-webpack-plugin`, certain options were deliberately excluded. For example, routes cannot use the `broadcastUpdate` option, since we would need to re-implement that functionality if we decided to move away from workbox in the future. However, we can of course add support for this property should the need arise.

Finally, I quickly updated the [hnpwa](https://github.com/mwistrand/dojo-2-hn/tree/cli-build-app-sw) example with the necessary [`.dojorc`](https://github.com/mwistrand/dojo-2-hn/blob/cli-build-app-sw/.dojorc#L17-L23) required to render the app offline. The two routes are the same, but the first is used for the deployed app, while the second is specific to the local, development environment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -363,6 +364,89 @@ However, in the following scenario the default locale will be used, although it
...
+`networkFirst` attempts to load a resource over the network, falling back on the cache if the request fails or times out. This is a useful strategy for assets that either change frequently or may change frequently (i.e., are not versioned). With the `cacheFirst` strategy, a resource is loaded from the cache unless it does not exist, in which case it is fetched over the network. `cacheFirst` is best for resources that change in frequently or can be cached for a long time (e.g., versioned assets). The `networkOnly` strategy forces the resource to always be retrieved over the network, and is useful for requests that have no offline equivalent. Finally, `staleWhileRevalidate` requests resources from both the cache and the network simulaneously. The cache is updated with each successful network response. This strategy is best for resources that do not need to be continuously up-to-date, like user avatars. However, when fetching third-party resources that do not send CORS headers, it is not possible to read the contents of the response or verify the status code. As such, it is possible that a bad response could be cached. In such cases, the `networkFirst` strategy may be a better fit.
```
Comment (dylans) [inline README.md]: This would probably be better as a list for the four strategies, or 4 separate paragraphs?

Comment (dylans) [inline README.md]: "is best for resources that change in frequently or can be cached for a long time" -> did you mean infrequently?

File: README.md [hunk 2]
```diff
@@ -363,6 +364,89 @@ However, in the following scenario the default locale will be used, although it
...
+| cacheName | `string` | Yes | The name of the cache to use for the route. Note that the `cachePrefix` is _not_ prepended to the cache name. Defaults to the main runtime cache (`${cachePrefix}-runtime-${domain}`). |
+| cacheableResponse | `object` | Yes | Uses HTTP status codes and or headers to determine whether a response can be cached. This object has two optional properties: `statuses` and `headers`. `statuses` is an array of HTTP status codes that should be considered valid for the cache. `headers` is an object of HTTP header and value pairs; at least one header must match for the response to be considered valid. Defaults to `{ statuses: [ 200 ] }` when the `strategy` is `'cacheFirst'`, and `{ statuses: [0, 200] }` when the `strategy` is either `networkFirst` or `staleWhileRevalidate`. |
+| expiration | `object` | Yes | Controls how the cache is invalidated. This object has two optional properties. `maxEntries` is the number of responses that can be cached at any given time. Once this max is exceeded, the oldest entry is removed. `maxAgeSeconds` is oldest a cached response can be in seconds before it is removed. |
```
Comment (dylans) [inline README.md]: "`maxAgeSeconds` is oldest a cached response can be in seconds before it is removed" perhaps better as "`maxAgeSeconds` is the oldest a cached response can be in seconds before it gets removed"

**Comment by mwistrand** [comment]
Thanks, @dylans. I updated the README to address your feedback.

**Comment by mwistrand** [comment]
I should note that workbox logs updates to the console. If that should be disabled, there is an option for that. Also, when providing a custom service worker, only that service worker file is copied to the output directory. So if there are additional `importScripts` that exist outside of the build pipeline, they will not be copied.

**Pull request merged**